crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1023,18 +1031,104 @@ impl<'db> TypeInferenceBuilder<'db> {
+    fn infer_context_expression(
+        &mut self,
+        context_expression: &ast::Expr,
+        context_expression_ty: Type<'db>,
+    ) -> Type<'db> {
+        let context_manager_ty = context_expression_ty.to_meta_type(self.db);
+
+        let enter_ty = context_manager_ty.member(self.db, "__enter__");
+        let exit_ty = context_manager_ty.member(self.db, "__exit__");
+
+        // TODO: The checks here should be simplified to checking if context manager implements the `contextlib.AbstractContextManager` protocol.
+        if enter_ty.is_unbound() && exit_ty.is_unbound() {
+            self.add_diagnostic(
+                context_expression.into(),
+                "invalid-context-manager",
+                format_args!(
+                    "Object of type {} cannot be used with `with` because it doesn't implement `__enter__` and `__exit__`",
+                    context_expression_ty.display(self.db)
+                ),
+            );
+            Type::Unknown
+        } else if enter_ty.is_unbound() {
+            self.add_diagnostic(
+                context_expression.into(),
+                "invalid-context-manager",
+                format_args!(
+                    "Object of type {} cannot be used with `with` because it doesn't implement `__enter__`",
+                    context_expression_ty.display(self.db)
+                ),
+            );
+            Type::Unknown
+        } else {
+            let target_ty = enter_ty
+                .call(self.db, &[context_expression_ty])
+                .return_ty_result(self.db, context_expression.into(), self)
+                .unwrap_or_else(|err| {
+                    self.add_diagnostic(
+                        context_expression.into(),
+                        "invalid-context-manager",
+                        format_args!("
+                            Object of type {context_expression} cannot be used with `with` because the method `__enter__` of type {enter_ty} is not callable",
+                            context_expression = context_expression_ty.display(self.db),
+                            enter_ty = enter_ty.display(self.db)
+                        ),
+                    );
+                    err.return_ty()
+                });
+
+            if exit_ty.is_unbound() {
+                self.add_diagnostic(
+                    context_expression.into(),
+                    "invalid-context-manager",
+                    format_args!(
+                        "Object of type {} cannot be used with `with` because it doesn't implement `__exit__`",
+                        context_expression_ty.display(self.db)
+                    ),
+                );
+            } else if exit_ty
+                .call(
+                    self.db,
+                    &[context_manager_ty, Type::None, Type::None, Type::None],
+                )
+                .return_ty_result(self.db, context_expression.into(), self)
```

Comment:
The return type here will matter to us later, because it will determine whether exceptions are caught by this `with` block. We could add a TODO for that but I don't think it's important that we do.